arangod/Scheduler/Scheduler.cpp

```diff
@@ -208,24 +208,41 @@ Scheduler::~Scheduler() {
 // do not pass callback by reference, might get deleted before execution
 void Scheduler::post(std::function<void()> const callback) {
   // increment number of queued and guard against exceptions
+  //  (this incQueued() manipulates the atomic _counters in a sequentially-consistent
+  //   manner.  isStopping() uses same atomic _counters)
   incQueued();
 
-  auto guardQueue = scopeGuard([this]() { decQueued(); });
+  // implies if _ioContext still valid (defense against shutdown races)
+  if (!isStopping()) {
+
+    auto guardQueue = scopeGuard([this]() { decQueued(); });
+
+    // capture without self, ioContext will not live longer than scheduler
+    _ioContext->post([this, callback]() {
+        // start working
+        JobGuard jobGuard(this);
+        jobGuard.work();
 
-  // capture without self, ioContext will not live longer than scheduler
-  _ioContext->post([this, callback]() {
-    // start working
-    JobGuard jobGuard(this);
-    jobGuard.work();
+        // reduce number of queued now
+        decQueued();
 
+        callback();
+      });
+
+    // no exception happened, cancel guard
+    guardQueue.cancel();
+  } else {
     // reduce number of queued now
     decQueued();
 
-    callback();
-  });
+    // this post is coming late in application shutdown,
+    //  might be essential ... process in-line after heavy complaining
+    LOG_TOPIC(WARN, Logger::THREADS)
+      << "Scheduler::post() called after isStopping() set.";
+    TRI_ASSERT(false);
 
-  // no exception happened, cancel guard
-  guardQueue.cancel();
+    callback();
```

Comment:
I believe JobGuard() is the incRunning()/decRunning() wrapper.  Yes?  @goedderz

Reply:
@goedderz Wait.  I am looking at the wrong "callback()".  Line numbers are not lining up correctly for me.

Reply:
@goedderz I am looking at the impact to Scheduler::shutdown() if we add the incRunning / decRunning to this "else" clause.  It seems to me that we get the same "keep shutdown from finishing" effect if we use the JobGuard object that manipulates the "working" count.  The advantage being that JobGuard will work in an exception situation too.  Would adding JobGuard lines from the lambda to this "else" clause satisfy your concerns?  I would put the constructor and work() call before the decQueued() call.

Reply:
Of course, the `decRunning()` would have to be added in a `TRI_DEFER` or `scopeGuard` to be exception-safe, too.

I notice that `SchedulerFeature::stop()` will only wait on `isRunning()`, not on `numWorking() > 0`, before calling `Scheduler::shutdown()`. Not saying that this is a problem, it's just a difference. I agree that `Scheduler::shutdown()` will work the same way when using a `JobGuard`.

The real problem in both cases is of course other threads rudely calling post after shutdown. And we cannot prevent that here either way. So that's fine with me.